Update filesystem tenancy documentation (#201)
stancl/tenancy-docs: source/docs/v3/tenancy-bootstrappers.blade.md

Note that you must use a cache store that supports tagging, e.g. Redis.

## Filesystem tenancy bootstrapper {#filesystem-tenancy-boostrapper}

> The filesystem bootstrapper makes your app's `storage_path()` and `asset()` helper and the `Storage` facade tenant-aware by modifying the paths they retrieve.

Reviewer: ```suggestion
The filesystem bootstrapper makes your app's `Storage` facade and the `storage_path()` and `asset()` helpers tenant-aware by modifying the paths they return.
```

> Note: If you want to bootstrap filesystem tenancy differently (e.g. provision an S3 bucket for each tenant), you can absolutely do that. Take a look at the package's bootstrappers to get an idea of how to write one yourself, and feel free to implement it any way you want.

### Storage path helper

> The bootstrapper suffixes the path retrieved by `storage_path()` to make the helper tenant-aware.

Reviewer: `returned` is probably better terminology

```suggestion
The bootstrapper suffixes the path returned by `storage_path()` to make the helper tenant-aware.
```

Logs will be saved in `storage/logs` regardless of any changes to `storage_path()` and regardless of the tenant.

### Storage facade

The bootstrapper also makes the `Storage` facade tenant-aware by suffixing the roots of disks listed in `config('tenancy.filesystem.disks')` and by overriding the disk roots in `config('tenancy.filesystem.root_override')` (disk root = the disk path used by the `Storage` facade).

The root of each disk listed in `config('tenancy.filesystem.disks')` will be suffixed. Doing that alone could cause unwanted behavior since Laravel does its own suffixing, so the filesystem config has the `root_override` section, which lets you override the disk roots **after** tenancy has been initialized:

```php
// Tenancy config (tenancy.filesystem.root_override)
// %storage_path% gets replaced by storage_path()'s output
// E.g. Storage::disk('local')->path('') will return "/$path_to_your_application/storage/tenant42/app"
// (Given a suffix_base of 'tenant' and a tenant with an ID of 42. Same as in the example above in the Storage path helper section)
'root_override' => [
    'local' => '%storage_path%/app/',
    'public' => '%storage_path%/app/public/',
],
```

> To make the tenant-aware `Storage` facade work with a custom disk, add the disk's name to `config('tenancy.filesystem.disks')` and override its root in `config('tenancy.filesystem.root_override')` as shown above.

Reviewer: Perhaps it's worth mentioning that root_overrides are mostly only relevant for local disks and not S3 (not 100% sure but I believe that's the case).

You could add a short paragraph saying that for S3, this is not necessary, and then include an example showing that `Storage::disk('s3')->path('foo.txt')` will return `/tenant42/foo.txt`

Author: Updated the paragraph https://github.com/stancl/tenancy-docs/pull/201/commits/9409f28393d31ff101b874bac725024cf732bb3d

### Assets

> The filesystem bootstrapper makes the `asset()` helper link to the files *of the currently initialized and identified tenant*. By default, the bootstrapper makes the helper output a URL pointing to the TenantAssetsController (`/tenancy/assets/...`), which returns a file response:

Reviewer: ```suggestion
The filesystem bootstrapper makes the `asset()` helper link to the files *of the current tenant*. By default, the bootstrapper makes the helper output a URL pointing to the TenantAssetsController (`/tenancy/assets/...`), which returns a file response:
```

```php
// TenantAssetsController
return response()->file(storage_path('app/public/' . $path));
```

The package expects the assets to be stored in your tenant's `app/public/` directory. For global assets (non-private assets shared among all tenants or JS/CSS assets), you may want to create a disk and use URLs from that disk instead. For example:

```php

> Storage::disk('app-public')->url('tenants/logos/' . tenant()->id . '.png');

Reviewer: The example does not follow the paragraph above it.

1. The paragraph talks about JS/CSS assets, but for those the `Storage` facade is generally never used. I'd remove that from the paragraph and instead add it to the paragraph below ("To access the global assets **such as JS/CSS assets**" — bold text indicates addition).
2. The paragraph says `assets shared among all tenants` but your example fetches a tenant-specific image. It'd be better to make the example something like `Storage::disk('branding')->url('header-logo.png')`

Author: https://github.com/stancl/tenancy-docs/pull/201/commits/92a70b7a4dae40e40c16471d293608de5d81c62d

```

To access the global assets, you can use `global_asset()` and `mix()`.

> The `asset()` helper's behavior can be changed by configuring the asset URL (`ASSET_URL` in your `.env`) – when the asset URL is not null, the bootstrapper will suffix the configured asset URL (the same way as `storage_path()` gets suffixed), and make the `asset()` helper output that.

Reviewer: I'd rather say that the behavior **changes** when the asset URL is set. Instead of saying that it "can be changed". Reason for this: sometimes the change happens without the user intentionally doing that, such as when using Vapor.

I'd also prefer the language of `is set` than `is not null`

Author: https://github.com/stancl/tenancy-docs/pull/201/commits/bdb6a4e9c7fd6db381a9eaf86fd432e5e1b5d290

> Configuring the asset URL (`ASSET_URL` in your `.env`) changes the `asset()` helper's behavior – when the asset URL is set, the bootstrapper will suffix the configured asset URL (the same way `storage_path()` gets suffixed), and make the `asset()` helper output that.

Reviewer: Maybe add something like `instead of returning a path to the TenantAssetsController`. So that the value returned when `ASSET_URL` is contrasted with what is returned normally.

Author: https://github.com/stancl/tenancy-docs/pull/201/commits/e85f708383576c8a06a305f4264de76727c0dd89

> You can disable tenancy of `asset()` in the config (`tenancy.filesystem.asset_helper_tenancy`) and explicitly use `tenant_asset()` instead. `tenant_asset()` returns your website's URL suffixed by `tenancy/assets/$nameOfYourAsset` (e.g. `https://your-site.test/tenancy/assets/asset-name.txt`). You may want to do that if you're facing issues using a package that utilizes `asset()` inside the tenant app.

Reviewer: 1. I'd change `returns your website's URL suffixed by` to `**always** returns a path to the TenantAssetController: tenant_asset('foo.txt') returns your-site.com/tenancy/assets/foo.txt` (and wrap the URL in `, I couldn't do that here due to formatting limitations)
2. Could you explain what you mean by the `You may want to do that if you're facing issues using a package that ...` point?

Author: > Could you explain what you mean by the You may want to do that if you're facing issues using a package that ... point?

If I'm using a package that uses `asset()` e.g. to get non-tenant-specific assets, the package won't get the asset paths correctly because of tenancy, so I'd want to disable `asset()` tenancy and use `tenant_asset()` instead in the app

Author: https://github.com/stancl/tenancy-docs/pull/201/commits/2aa7e44d4368152adce590c4e8403dd54fdb5eb7

Reviewer: Re: 2) I see, that makes sense. When I was reading this yesterday I read only the second part after writing the review for the first sentence, so it seemed that this only says that you can use `tenant_asset()` if you're using 3rd party packages, but when you read the paragraph in full it does make sense since it's in reference to disabling the asset helper tenancy.

You can mark this as resolved 👍🏻

Note that you must use a cache store that supports tagging, e.g. Redis.

## Filesystem tenancy bootstrapper {#filesystem-tenancy-boostrapper}
The filesystem bootstrapper makes your app's `storage_path()` and `asset()` helper and the `Storage` facade tenant-aware by modifying the paths they retrieve.

> Note: If you want to bootstrap filesystem tenancy differently (e.g. provision an S3 bucket for each tenant), you can absolutely do that. Take a look at the package's bootstrappers to get an idea of how to write one yourself, and feel free to implement it any way you want.

### Storage path helper

The bootstrapper suffixes the path retrieved by `storage_path()` to make the helper tenant-aware.

- The suffix is `config('tenancy.filesystem.suffix_base') . $tenantKey`

> - The suffix base is 'tenant' by default, but feel free to change it in the tenancy config

Reviewer: ```suggestion
- The suffix_base is 'tenant' by default, but feel free to change it in the tenancy config
```
Probably more understandable in reference to the point above

> - After suffixing the `storage_path()` helper, it returns `"/$path_to_your_application/storage/$suffix/"`

Reviewer: This one seems a bit confusing, since `$suffix` here is the suffix base + tenant id, but the readers may not get that without seeing an example. So perhaps we could replace this by an example with a random tenant id and the full generated path?

Author: That might, be better, I'll update it and we'll see

Author: Added the example, but I'd keep the `config('tenancy....'). $tenantKey` because that can make things a bit clearer.

Reviewer: Looks like this isn't fully resolved. I meant that `/$path_to_your_application/storage/$suffix/` should be `/$path_to_your_application/storage/tenant42/` given the example values you added. The examples are great but I'd make this change as well to make the final string in the example complete, rather than having variables in it (with the exception of `$path_to_your_application` which we of course don't know).

Reviewer: Created a suggestion that you can just commit and then you can resolve this https://github.com/stancl/tenancy-docs/pull/201/files#r979141050

Since `storage_path()` will be suffixed, your folder structure will look like this:

![The folder structure](https://i.imgur.com/GAXQOnN.png)

Logs will be saved in `storage/logs` regardless of any changes to `storage_path()` and regardless of the tenant.

### Storage facade

> The bootstrapper also makes the `Storage` facade tenant-aware by suffixing the roots of disks listed in `config('tenancy.filesystem.disks')` and by overriding the roots in `config('tenancy.filesystem.root_override')`.

Reviewer: > and by overriding the roots in `config('tenancy.filesystem.root_override')`

This part is a bit hard to understand

Author: How would you make it easier to understand?? I thought adding 'disk' before 'roots' might make it clearer, but I think you might mean somethings else

Author: Updated this a bit

Reviewer: I think what I meant here is that the sentence is complex and in general the `root_override` logic isn't something people intuitively have in their mind, so it's better to explain this in a way that's clear and understandable.

Author: Yeah, that would be the best. I [added some info](https://github.com/stancl/tenancy-docs/pull/201/files#diff-8a5121872d845abceb00e66b26a66c06700c839e289cd160af08612cc4c71cdeR51), hope that helps to make the section understandable

```php
// Tenancy config (tenancy.filesystem.root_override)
// %storage_path% gets replaced by storage_path()'s output
// E.g. Storage::disk('local')->path('') will return "/$path_to_your_application/storage/$suffix/app"
'root_override' => [
    'local' => '%storage_path%/app/',
    'public' => '%storage_path%/app/public/',
],
```

To make the tenant-aware `Storage` facade work with a custom disk, add the disk's name to `config('tenancy.filesystem.disks')` and override its root in `config('tenancy.filesystem.root_override')` as shown above.

### Assets

The bootstrapper modifies the links retrieved by the `asset()` helper, so they link to the files *of the currently initialized and identified tenant*.

Before using the `asset()` helper, make sure to [assign the identification middleware you're using in your app to TenantAssetsController's `$tenancyMiddleware`]({{ $page->link('configuration#static-properties') }}):

```php
// TenancyServiceProvider (don't forget to import the classes)

public function boot()
{
    // Update the middleware used by the asset controller
    TenantAssetsController::$tenancyMiddleware = InitializeTenancyByDomainOrSubdomain::class;
}
```

> > Note: You can disable tenancy of `asset()` in the config (`tenancy.filesystem.asset_helper_tenancy`) and explicitly use `tenant_asset()` instead. You may want to do that if you're facing issues using a package that utilizes `asset()` inside the tenant app.

Reviewer: Do `asset()` and `tenant_asset()` work interchangeably? When asset helper tenancy is enabled, that is.

Author: I just read @abrardev99's comments about the helpers, and my bad for not trying this out myself, I was under the impression that they had the same output when asset helper tenancy was enabled. 

Responded on BC: https://3.basecamp.com/5170965/buckets/23651018/todos/4796489363#__recording_5347808448

Reviewer: > I was under the impression that they had the same output when asset helper tenancy was enabled.

Same, that's why I was confused about the BC thread.

Author: Should I update this now, or should we wait for the `tenant_asset()` update?

Reviewer: You're writing v3 docs, v4 changes don't have an impact on this. Or what do you mean by update?

Author: > Or what do you mean by update?

I meant making `tenant_asset()` work as we expected it to work. You're implying that'll be done in v4, so I'll update the docs to match how it works in v3.

Author: Updated

Reviewer: We won't be making any changes to `tenant_asset()` in v3, way too many breaking changes could come out of that.

>
> For non-tenant-specific assets (assets shared among all tenants or JS/CSS assets), you can use `global_asset()` and `mix()`.

If `config('app.asset_url')` has been set, the bootstrapper suffixes the configured asset URL the same way as `storage_path()` (useful if you're using Laravel Vapor or similar).

If `config('app.asset_url')` is null (as it is by default), `asset()` will return a URL pointing to the TenantAssetsController (`/tenancy/assets/...`) which returns a file response:

```php
// TenantAssetsController
return response()->file(storage_path('app/public/' . $path));
```

> When `config('app.asset_url')` is not set, **you need to store the assets in your tenant's `app/public/` directory**. So for non-private (global) assets, you may want to create a disk and use URLs from that disk instead (don't add that disk to `config(tenancy.filesystem.disks)`). For example:

Reviewer: The `is not set` here is slightly confusing because the above paragraph talks about it being null, so it's not clear if they're the same things or not.

Author: I see, it should be `null`

Note that you must use a cache store that supports tagging, e.g. Redis.

## Filesystem tenancy bootstrapper {#filesystem-tenancy-boostrapper}
The filesystem bootstrapper makes your app's `storage_path()` and `asset()` helper and the `Storage` facade tenant-aware by modifying the paths they retrieve.

> Note: If you want to bootstrap filesystem tenancy differently (e.g. provision an S3 bucket for each tenant), you can absolutely do that. Take a look at the package's bootstrappers to get an idea of how to write one yourself, and feel free to implement it any way you want.

### Storage path helper

The bootstrapper suffixes the path retrieved by `storage_path()` to make the helper tenant-aware.

- The suffix is built by appending the tenant key to your `suffix_base`. The `suffix_base` is `tenant` by default, but feel free to change it in the `tenancy.filesystem` config. For example, the suffix will be `tenant42` if the tenant's key is `42` and the `suffix_base` is `tenant`.

> - After the suffixing, `storage_path()` helper returns `"/$path_to_your_application/storage/$suffix/"`

Reviewer: ```suggestion
- After the suffixing, `storage_path()` helper returns `"/$path_to_your_application/storage/tenant42/"`
```
Related to my comment in the other review

Since `storage_path()` will be suffixed, your folder structure will look like this:

> ![The folder structure](https://i.imgur.com/GAXQOnN.png)

Reviewer: Could you move the image to our repo (https://github.com/stancl/tenancy-docs/tree/master/source/assets/images) and link to it like this https://github.com/stancl/tenancy-docs/blob/master/source/docs/v3/early-identification.blade.md?plain=1#L71

It will not display in this PR until it's merged, but after that it will work and it will be a reliable part of our repo.

Author: The same imgur link was used for v1 and v2 docs too, so I changed the links there too.

Logs will be saved in `storage/logs` regardless of any changes to `storage_path()` and regardless of the tenant.

### Storage facade

The bootstrapper also makes the `Storage` facade tenant-aware by suffixing the roots of disks listed in `config('tenancy.filesystem.disks')` and by overriding the disk roots in `config('tenancy.filesystem.root_override')` (disk root = the disk path used by the `Storage` facade).

The root of each disk listed in `config('tenancy.filesystem.disks')` will be suffixed. Doing that alone could cause unwanted behavior since Laravel does its own suffixing, so the filesystem config has the `root_override` section, which lets you override the disk roots **after** tenancy has been initialized:

```php
// Tenancy config (tenancy.filesystem.root_override)
// %storage_path% gets replaced by storage_path()'s output

> // E.g. Storage::disk('local')->path('') will return "/$path_to_your_application/storage/$suffix/app"

Reviewer: Maybe make the change here as well? For clarity.
```suggestion
// E.g. Storage::disk('local')->path('') will return "/$path_to_your_application/storage/tenant42/app"
// (Given a suffix_base of 'tenant' and a tenant with an ID of 42. Same as in the example above in the Storage path helper section)
```

### Assets

The bootstrapper modifies the links retrieved by the `asset()` helper, so they link to the files *of the currently initialized and identified tenant*.

Before using the `asset()` helper, make sure to [assign the identification middleware you're using in your app to TenantAssetsController's `$tenancyMiddleware`]({{ $page->link('configuration#static-properties') }}):

```php
// TenancyServiceProvider (don't forget to import the classes)

public function boot()
{
    // Update the middleware used by the asset controller
    TenantAssetsController::$tenancyMiddleware = InitializeTenancyByDomainOrSubdomain::class;
}
```

> Note: You can disable tenancy of `asset()` in the config (`tenancy.filesystem.asset_helper_tenancy`) and explicitly use `tenant_asset()` instead. But keep in mind that `tenant_asset()` does not respect the asset URL you set in `config('app.asset_url')` – the helper will always return your website's url suffixed by `tenancy/assets/name-of-your-asset.foo` (e.g. `https://your-site.test/tenancy/assets/asset-name.txt`).
>
> You may want to do that if you're facing issues using a package that utilizes `asset()` inside the tenant app.
>

> > For non-tenant-specific assets (assets shared among all tenants or JS/CSS assets), you can use `global_asset()` and `mix()`.

Reviewer: Perhaps mention that this is an alternative to disabling asset helper tenancy?

So change it to something like:
```suggestion
> As an alternative to disabling `asset_helper_tenancy`, you may use `global_asset()` and `mix()` for non-tenant-specific assets (assets shared among all tenants or JS/CSS assets).
```

Author: I think saying that it's an alternative makes it look like we're suggesting that disabling the asset helper tenancy is the primary way to go about accessing the non-tenant-specific assets. Maybe the above note about disabling asset_helper_tenancy would be enough?

Reviewer: True, you can ignore this suggestion if you do the restructure I mentioned.

If `config('app.asset_url')` has been set, the bootstrapper suffixes the configured asset URL the same way as `storage_path()` (useful if you're using Laravel Vapor or similar).

> If `config('app.asset_url')` is null (as it is by default), `asset()` will return a URL pointing to the TenantAssetsController (`/tenancy/assets/...`) which returns a file response:

Reviewer: ```suggestion
If `config('app.asset_url')` is null (as it is by default), `asset()` will return the same output as `tenant_asset()` — a URL pointing to the TenantAssetsController (`/tenancy/assets/...`) which returns a file response:
```

```php
// TenantAssetsController
return response()->file(storage_path('app/public/' . $path));
```

> When `config('app.asset_url')` is null, **you need to store the assets in your tenant's `app/public/` directory**. So for non-private (global) assets, you may want to create a disk and use URLs from that disk instead (don't add that disk to `config(tenancy.filesystem.disks)`). For example:

Reviewer: This section is still a bit messy:
1. This paragraph directly follows another paragraph that starts basically the same: `If `config('app.asset_url')` is null`. It *feels* like it's a typo because the flow of the page suggests that this paragraph would be put in contrast against the previous paragraph, but they're talking about the same thing. When you read the section for the second time, it does make sense, but on first read it seems confusing.
2. There's a lot of text explaining edge cases or configuration *before* the actual behavior is explained.

I'd restructure this entire section. Start with what the package does out of the box, then talk about what happens when `ASSET_URL` is **not** null, then talk about global assets, then mention disabling asset helper tenancy, and then explain using the `tenant_asset()` helper.

The last few sections can be swapped depending on what you feel is best, but I'd definitely start with explaining what is being done out of the box, then explaining the **common** complexities that people may have to deal with (global assets, `ASSET_URL` being set by Vapor), and only then talk about the stuff you have in the note at the top. Maybe `ASSET_URL` can be at the very end since that's a Vapor-specific complexity, and much less relevant than something like global assets which pretty much everyone has to deal with.

I'd also first explain what the helper does, and only then mention the `TenantAssetsController::$tenancyMiddleware`. We first want to tell people **what this section is about**, and only then talk about all of the nuances and complexities of it.

Author: Thanks for the guidance, I restructured the section https://github.com/stancl/tenancy-docs/pull/201/commits/3263e2bb50a2d63028ea431aef88c7c0a00170b7
